[Don't merge][GSoC]Added ArrayComprehension class

kangzhiq opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Added ArrayComprehension class to generate a list comprehension which can accept symbols

#### Other comments
The implementation is not yet finished. 
PR related to GSoC 2019

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* tensor
  * Added a new array object ArrayComprehension for list comprehension.
<!-- END RELEASE NOTES -->


TODO list:

- [x] add unit tests
- [x] add doctests in the documentation string.

Commit:
Initial ArrayComprehension class

- Added constructor, doit()
- _expand_aray() to be completed

Commit:
Update _init_.py to include the new file

Upabjojr commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,36 @@
+
+    def _expand_array(self, limits):
+        print(limits)
+        return 0
```
missing new line, tests won't pass.

kangzhiq replied:
Thank you for reviewing! I will correct the coding convention errors in next commit.

Upabjojr commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,36 @@
+                return self
+            else:
+                self.expr = self._expand_array([inf, sup])
+
```
maximum one newline inside indented namespaces.

czgdp1807 commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,36 @@
+    def __new__(cls, expr, *values, **assumptions):
+        if any(len(l) != 3 or None for l in values):
+            raise ValueError('ArrayComprehension requires values lower and upper bound'
+                              'for the expression')
```
give a space between `'` and `for`  like,  ' for the expression')

kangzhiq replied:
I am sorry that I didn't see your review, I will uodate it in next commit. Thanks a lot!!

czgdp1807 commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,36 @@
+
+
+    def _expand_array(self, limits):
+        print(limits)
```
shouldn't it be `return limits`?

kangzhiq replied:
Yes, it is a fast test to verifier the output. I have update the implememntation of _extand_array, please feel free to review it. :-)

Upabjojr replied:
@kangzhiq are you using a debugger? You shouldn't use `print` to inspect variables. I suggest to use PyCharm to inspect the variables while running the code.

czgdp1807 commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,36 @@
+            # Array will not ne expanded if there is a symbolic dimension
+            if any(i.is_symbol for i in [inf, sup]):
+                return self
+            else:
```
`else` not needed. The control will automatically skip the `else` part if the `if` part is executed. In addition, using `not i.is_number` or `i.atoms(Symbol)` would be better otherwise, expressions like `N + 1` will go for expansion rather than being unevaluated. You can write it as,
```
if Basic(inf, sup).atoms(Symbol):
    return self
self.expr = self._expand_array([inf, sup])
```

kangzhiq replied:
That is a very valuable advice! Thank you! I will update it as you suggested!

Commit:
_expand_array() completed

- Completed first version of _expand_array
_ Correct the coding convention errors

Commit:
Modified symbol check condition and passed test

czgdp1807 commented:
Add the test in `sympy.core.tests.test_args`.

czgdp1807 commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,69 @@
+                arr[index] = arr[index].subs(var, val)
+            return arr.tolist()
+
+        # Recursive function to perform subs at every variable according to its boundary
```
I think `f` should be having iterative logic as `recursion` is expensive than `iteration` in Python as there is always an allocation of new stack frame at each call leading to an overhead.
I believe that there is a possibility of converting this recursive logic into an iterative one. At least try doing.

kangzhiq replied:
Yes, you are right! I have an idea to implement it in an iterative way. I will update it asap. Thank you for reviewing!

kangzhiq replied:
> so better avoid them.

I just implemented a ew version with only iterations. I will commit it.

kangzhiq replied:
I am afraid that `itertools` might not be very helpful, since it is an iteration over a list with Symbol. I will still thinking about it but after a first try, I didn't find a way to use it.

Commit:
Merge branch 'master' of https://github.com/sympy/sympy into Added_ArrayComprehension_Class

Commit:
Added test for ArrayComprehension

- added test in test_args
- added test file for ArrayComprehension

Commit:
Added more tests for unimplemented functionalities

sylee957 commented:
I can see how the implementation looks like, after tests had been added.
I think that the term comprehension can be ambiguous, but from the python [doc](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions)


> A list comprehension consists of brackets containing an expression followed by a for clause, then zero or more for or if clauses.

And I don't know if `ArrayComprehension` is using `for` loops.

It can be more correspondent to the [Table](https://reference.wolfram.com/language/ref/Table.html) in Mathematica, though.

kangzhiq commented:
@sylee957  Thank you for reviewing!
> It can be more correspondent to the [Table](https://reference.wolfram.com/language/ref/Table.html) in Mathematica, though.

That Table module is exactly what I am trying to reproduce here. And yes, ArrayComprehension is using `for` loops.
For example `ArrayComprehension(i, (i, 1, 5))` is equivalent to `[i for i in range(1, 5)]`, except that the built-in list comprehension of Python doesn't accept Symbol, e.g. `[i for i in range(1, n)]` where `n` is a Symbol which value is unknown. So ArrayComprehension will be useful in this specific case.

Upabjojr commented on sympy/tensor/array/tests/test_array_comprehension.py:
```diff
@@ -0,0 +1,14 @@
+    assert b.subs(j, 3) == [1, 2, 3]
+    assert c.doit == [j, 2*j, 3*j]
+    c.add_bound((j, 2, 4))
+    assert c.doit() == [[2, 3, 4], [4, 6, 8], [6, 9, 12]]
```
We need some tests for larger dimensions (e.g. 4, 5 variables).

Upabjojr commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,69 @@
+        return arr
+
+    # Add/replace a boudary of variable
+    def add_bound(self, bound):
```
SymPy methods are immutable. No edits after class constructor, otherwise bad things may happen (i.e. problems with hashing, term rewriter, etc...).

kangzhiq replied:
I didn't know that. Thank for telling me :-)

Upabjojr commented on sympy/tensor/array/tests/test_array_comprehension.py:
```diff
@@ -0,0 +1,14 @@
+    assert isinstance(b.doit(), ArrayComprehension)
+    assert b.subs(j, 3) == [1, 2, 3]
+    assert c.doit == [j, 2*j, 3*j]
+    c.add_bound((j, 2, 4))
```
`c` cannot be modified after construction. This is a SymPy standard.

Upabjojr commented on sympy/tensor/array/tests/test_array_comprehension.py:
```diff
@@ -0,0 +1,14 @@
+    c = ArrayComprehension(i*j, (i, 1, 3))
+    assert a.doit() == [[2, 3, 4], [4, 6, 8], [6, 9, 12]]
+    assert isinstance(b.doit(), ArrayComprehension)
+    assert b.subs(j, 3) == [1, 2, 3]
```
why is `.subs` triggering the evaluation? I suppose it should create `ArrayComprehension(i, (i, 1, 3))`.

kangzhiq replied:
I was thinking that `.subs ` try do expand the array (calling a doit() inside) if all variable has a specific bound or value. It should be `[3, 6, 9]` instead, I made a mistake.

Upabjojr commented on sympy/tensor/array/tests/test_array_comprehension.py:
```diff
@@ -0,0 +1,14 @@
+    a = ArrayComprehension(i*j, (i, 1, 3), (j, 2, 4))
+    b = ArrayComprehension(i, (i, 1, j))
+    c = ArrayComprehension(i*j, (i, 1, 3))
+    assert a.doit() == [[2, 3, 4], [4, 6, 8], [6, 9, 12]]
```
can you also add `assert isinstance(a.doit(), ImmutableDenseNDimArray)`?

kangzhiq replied:
Do does it mean that ArrayComprehension should inherit from ImmutableDenseNDimArray?

Upabjojr commented:
Also, `ArrayComprehension( ... ).free_symbols` should not return the symbols being expanded in.

kangzhiq commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,69 @@
+
+class ArrayComprehension(Basic):
+    """
+    Generate a list comprehension
```
I didn't know well itertools very well. I will have a look at it!
doctests will be added imediately.

Commit:
Implemented functions and added tests

- Implemented _expand_array with iteration
- Implemented _check_bounds_validity
- Added doctest
- Added test in test_array_comprehension

Upabjojr commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,90 @@
+
+    >>> from sympy.tensor.array import ArrayComprehension
+    >>> from sympy.abc import i, j, k
+    >>> a = ArrayComprehension(10*i+j, (i, 1, 4), (j, 1, 3))
```
show that `a` has not been evaluated.

Upabjojr commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,90 @@
+        bounds = sympify(bounds)
+        for var, inf, sup in bounds:
+            if var not in expr.free_symbols:
+                raise ValueError('Varialbe {} does not exist in expression'.format(var))
```
why? what about generating constant arrays? I don't see any problem with `ArrayComprehension(1, (i, 1, N))`.

Also, add a test to check that this is possible.

Upabjojr commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,90 @@
+            raise ValueError('ArrayComprehension requires values lower and upper bound'
+                              ' for the expression')
+        cls.default_assumptions = assumptions
+        obj = Expr.__new__(cls, **assumptions)
```
the arguments `expr` and `bounds` need to be passed to the superclass in the canonical form. That is, you should be able to rebuild the same object with the arguments you pass to the superclass.

kangzhiq replied:
@Upabjojr Thank you for reviewing!
Now I understand better the structure of a new class.

Upabjojr commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,90 @@
+                              ' for the expression')
+        cls.default_assumptions = assumptions
+        obj = Expr.__new__(cls, **assumptions)
+        obj.expr = expr
```
`obj._expr = expr`, then add a property for `expr`.

Upabjojr commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,90 @@
+        cls.default_assumptions = assumptions
+        obj = Expr.__new__(cls, **assumptions)
+        obj.expr = expr
+        obj.bounds = cls._check_bounds_validity(expr, bounds)
```
same as above

Upabjojr commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,90 @@
+        obj.bounds = cls._check_bounds_validity(expr, bounds)
+        arglist = [expr]
+        arglist.extend(obj.bounds)
+        obj._args = tuple(arglist)
```
WRONG. This is handled by the superclass in `Expr.__new__`, don't modify `_args`.

kangzhiq replied:
@Upabjojr 
Yes, I just modified the code. By the way, I was following the code of ExprWithLimits as an example, [Code](https://github.com/sympy/sympy/blob/c50643a49811e9fe2f4851adff4313ad46f7325e/sympy/concrete/expr_with_limits.py#L164-L171). So should these codes be standardized as well?

Upabjojr commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,90 @@
+    def doit(self):
+        expr = self.expr
+        for index, value in enumerate(self.bounds):
+            var, inf, sup = map(sympify, value)
```
maybe check that `inf` is zero. We don't support offsetting the indices of arrays yet.

kangzhiq replied:
@Upabjojr What would be the problem please? 
In fact, one thing matters about `inf` and `sup` is their difference. So inf can be positif or negatif, as long as `sup - inf` is positif. 
if you are thinking about the case where `sup == inf`, there will not be a problem because the upper bound is included. It is `]inf, sup]` :-)

Upabjojr replied:
I was thinking of a different thing. But OK, let's keep it like this.

Upabjojr commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,90 @@
+            if var not in expr.free_symbols:
+                raise ValueError('Varialbe {} does not exist in expression'.format(var))
+            if any(not isinstance(i, (Integer, Symbol)) for i in [inf, sup]):
+                raise TypeError('Bounds should be an Integer or a Symbol')
```
why? I don't see any problem with expressions as bounds. Maybe also add a test `ArrayComprehension(..., (i, 0, N+1))`.

kangzhiq replied:
@Upabjojr Infact, I was trying to eliminate the cases where inf and sup are other than expressions, e.g. ArrayComprehension(k, (i, 1, [2, 3, 3, 3])) or ArrayComprehension(k, (i, 1, ArrayComprehension(k, (i, 1, 5)))), this does not really make sense. But I can't find a good solution for now.

Upabjojr commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,90 @@
+            if any(not isinstance(i, (Integer, Symbol)) for i in [inf, sup]):
+                raise TypeError('Bounds should be an Integer or a Symbol')
+            if isinstance(inf, Integer) and isinstance(sup, Integer) and inf > sup:
+                raise ValueError('Lower bound should be inferior to upper bound')
```
`if (inf > sup) == True:`

kangzhiq replied:
@Upabjojr Could you please tell me the advantage of doing so? Or it is a coding convention?

Upabjojr replied:
you don't need to check that `inf` and `sup` are integers. The `>` sign creates a `StrictGreaterThan` object if it cannot determine the truth immediately. If you are sure the inequality is always true, it returns `True`.

Please remove `isinstance(..., Integer)`.

kangzhiq replied:
Yes, you are right. As well as a bound with two expressions like `N+1` and `N+5`， it should be valide.

Upabjojr commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,90 @@
+
+    # Substitute the variable with a value, so that the symbolic dimension can be expanded as well
+    def subs(self, var, val):
+        return 0
```
Remove this. This is handled by `sympy/core` and should not be overwritten.

Upabjojr commented on sympy/tensor/array/tests/test_array_comprehension.py:
```diff
@@ -0,0 +1,23 @@
+    assert a.doit().tolist() == [[2, 3, 4], [4, 6, 8], [6, 9, 12]]
+    assert isinstance(b.doit(), ArrayComprehension)
+    assert isinstance(a.doit(), ImmutableDenseNDimArray)
+    #assert b.subs(j, 3) == [3, 6, 9]
```
You can still test `subs`.

Upabjojr commented on sympy/tensor/array/tests/test_array_comprehension.py:
```diff
@@ -0,0 +1,23 @@
+                                 [[[5, 6, 7, 8, 9], [6, 7, 8, 9, 10], [7, 8, 9, 10, 11], [8, 9, 10, 11, 12]],
+                                  [[6, 7, 8, 9, 10], [7, 8, 9, 10, 11], [8, 9, 10, 11, 12], [9, 10, 11, 12, 13]],
+                                  [[7, 8, 9, 10, 11], [8, 9, 10, 11, 12], [9, 10, 11, 12, 13], [10, 11, 12, 13, 14]]]]
+    raises(TypeError, lambda:ArrayComprehension(i*j, (i, 1, 3), (j, 2, [1, 3, 2])))
```
space after `lambda:`

Commit:
Ameliorated functions and added test

- Ameliorated the value check of bounds
- Ameliorated the constructor
- Added properties and tests
- Implemented generation of constant arrays
- Added tests to ArrayComprehension

Upabjojr commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,121 @@
+        if any(len(l) != 3 or None for l in bounds):
+            raise ValueError('ArrayComprehension requires values lower and upper bound'
+                              ' for the expression')
+        cls.default_assumptions = assumptions
```
let this be handled by the superclass constructor.

Upabjojr commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,121 @@
+        >>> a.expr
+        10*i + j
+        """
+        return self._args[0]
```
we have `self._expr` now

Upabjojr commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,121 @@
+        return self._args[0]
+
+    @property
+    def bounds(self):
```
is this the same terminology used in `Sum` and `Product`? I remember a different name, can you check?

kangzhiq replied:
In `Sum` and `product`,  `function` ans `symbols` are used as parameters of construtor which correspond to `expr` and `bounds`. But then `symbols` is transformed `limits`. Since there is not a standard naming convention, I created the name by my own.
Besides, there are not explicit properties in `Sum` to make these parameters accesible. In `product`， there is only one for `function`, but it is renamed as `term`. 
In this case, what terminology do you think I should keep?

Upabjojr replied:
I think there is a standard terminology. There is a superclass implementing it.

kangzhiq replied:
@Upabjojr 
I think what you are refering to is the `ExprWithLimits` class. And yes, it defines many properties like `function`, `limits`, `variables`, `bound_symbols`, `free_symbols`, `is_number`. Do you prefer folowing these examples?

Upabjojr replied:
Maybe yes, if no complications arise I'm +1 with that. You can do this here or in another PR.

Upabjojr replied:
Maybe just make sure the names correspond with `ExprWithLimits`, I want this PR to be merged as soon as possible. Subclassing can be done in the future.

Upabjojr commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -0,0 +1,121 @@
+from sympy.core.compatibility import Iterable
+from sympy.core.numbers import Integer
+
+class ArrayComprehension(Basic):
```
This class is missing the `.shape` property, and the `.rank()` function as well. It should be compatible with the arrays.

Upabjojr commented on sympy/tensor/array/tests/test_array_comprehension.py:
```diff
@@ -0,0 +1,27 @@
+    assert isinstance(b.doit(), ArrayComprehension)
+    assert isinstance(a.doit(), ImmutableDenseNDimArray)
+    assert b.subs(j, 3) == ArrayComprehension(i, (i, 1, 4))
+    assert c.free_symbols == {i, j, k, l}
```
`c.free_symbols == set([])`. There are no free symbols in `c` as they are variables to be expanded.

You should subtract the iteration variables from `.expr.free_symbols`, then add the free symbols in the bounds.

Upabjojr replied:
For example:

```python
>>> ArrayComprehension(a*i, (i, k, N)).free_symbols
{a, k, N}
```
where `i` is removed because it won't exist anymore after calling `.doit()`.

czgdp1807 replied:
Will this implementation allow multivariate lists like,
```
>>> A = ArrayComprehension(i + j, (i, ki, Ni), (j, kj, Nj))
>>> A.free_symbols
{kj, Nj, ki, Ni}
```

Upabjojr replied:
Yes,  only `i` and `j` will be removed from free symbols in that example.

Commit:
Added free_symbols and tests

- Added property free_symbols and tests
- Added bound value check and test
- Ameliorated expr bounds properties

czgdp1807 commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -87,6 +111,8 @@ def _check_bounds_validity(cls, expr, bounds):
             if isinstance(inf, Integer) and isinstance(sup, Integer) and (inf > sup) == True:
                 raise ValueError('Lower bound should be inferior to upper bound')
+            if var in inf.free_symbols or var in sup.free_symbols:
+                raise ValueError('Variable should not be part of its bounds')
```
I think that's quite intuitive to have variable symbols to be replaced rather than constants. Moreover, instead of allowing only `Symbol`, expressions containing `Symbol` should be allowed. Like `(n + 1, 4, 8)`.

Upabjojr replied:
The variable needs to be a symbol, of course. `inf` and `sup` can be any expression.

> Moreover, instead of allowing only Symbol, expressions containing Symbol should be allowed. Like (n + 1, 4, 8).

No, use `(n, 4+1, 8+1)`. First variable needs to be a symbol.

Have you ever trying a list comprehension like this:
```python
[n for n+1 in range(4, 8)]
```
I don't think this should be allowed.

Commit:
Added properties to ArrayComprehehnsion

- Added .shape and tests
- Added .rank() and tests
- Added len() and tests

kangzhiq commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -103,18 +107,88 @@ def free_symbols(self):
+        12
+        """
+        if len(self._loop_size.free_symbols) != 0:
+            raise ValueError('Symbolic length is not supported')
```
Is it possible to find a bypass so that we can return an `expression` rather than `integer` for `len()`?
Or maybe another function/property like `size` to return this length of the expanded array.

czgdp1807 replied:
AFAIK, `__len__` always returns an integer, so recommended is using `size`. Something similar has been done in `Range` in `sympy.sets.fancysets`.

Upabjojr replied:
what about 
```python
return int(product([sup-inf for var, inf, sup in self.limits]))
```
?

kangzhiq replied:
I am sorry but I don't really understand how product will help in this case. Do you mean `itertool.product`? In fact, the problem is when we have an array like
```
>>> a = ArrayComprehension(10*i + j, (i, 1, 4), (j, 1, k + 3))
>>> len(a)
4 * (k + 3)
```
it is supposed to have an expression, but len() accepts only integer as return type.
For your information, a bound like `(i, j, j+4)` will return 4 as length so there is not problem if the difference is an `Integer`.

jksuom commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -40,12 +40,15 @@ def __new__(cls, expr, *bounds, **assumptions):
     def expr(self):
         """
-        Return the expression that will be expanded to an array
+        Returns the expression that will be expanded to an array
```
From PEP 257:
> The docstring is a phrase ending in a period. It prescribes the function or method's effect as a command ("Do this", "Return that"), not as a description; e.g. don't write "Returns the pathname ...".

Commit:
Changed to standard naming convention

- Follewed example of ExprWithLimits
- Changed expr to function
- Changed bounds to limits
- Added variables,bouns_symbols, is_numeric properties and tests
_ Ameliorated doit() and doc quality

Upabjojr commented:
OK, merged.
